keps/582-preempt-based-on-flavor-order/README.md

```diff
@@ -0,0 +1,330 @@
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Cluster Queue API
+
+We extend the Cluster Queue API to introduce the new fields: flavorFungibility to opt-in and configure the new behavior.
+
+```
+type FlavorFungibility string
+
+const (
+	BeforeNextFlavor FlavorFungibility = "BeforeNextFlavor"
+	AfterAllFlavors  FlavorFungibility = "AfterAllFlavors"
+)
+
+// ClusterQueuePreemption contains policies to preempt Workloads from this
+// ClusterQueue or the ClusterQueue's cohort.
+type ClusterQueuePreemption struct {
+	...
+
+	// +kubebuilder:default="AfterAllFlavors"
+	// +kubebuilder:validation:Enum=BeforeNextFlavor;AfterAllFlavors
+	FlavorFungibility FlavorFungibility `json:"flavorFungibility"`
+}
+```
+
+### Schduling Process in Kueue
+
+For each type of resource in each podSet, Kueue will traverse all resource groups and resource flavors to find a available flavor in present. And this will be the behavior for `AfterAllFlavors`.
+
+If flavorFungibility is BeforeNextFlavor, we will try to allocate podSet's resources in the first flavor in the resource group. Then we will try to preempt in these flavors if some resources not fit. If there are still some resource not fit, we will try the next flavor in the resource group until all resource types get allocated or there is no more resource flavors. 
+
+For example, if cluster queue has 2 resource groups and workload has 1 podSet as the following:
+
+```
+...
+  - coveredResources: ["cpu", "memory"]
+    flavors:
+    - name: "default-flavor1"
+      resources:
+      - name: "cpu"
+        nominalQuota: 3 
+      - name: "memory"
+        nominalQuota: 600Mi 
+    - name: "default-flavor2"
+      resources:
+      - name: "cpu"
+        nominalQuota: 3 
+      - name: "memory"
+        nominalQuota: 600Mi 
+  - coveredResources: ["gpu"]
+    flavors:
+    - name: "vendor1"
+      resources:
+      - name: "gpu"
+        nominalQuota: 9  
+    - name: "vendor2"
+      resources:
+      - name: "gpu"
+        nominalQuota: 9  
+---  
+...
+  podSets:
+  - count: 3
+    spec:
+      containers:
+      - ...
+        resources:
+          requests:
+            cpu: "1"
+            memory: 200Mi
+            gpu: 1
+```
+
+We will first try `default-flavor1` for cpu and memory resources. If `default-flavor1` doesn't fit, we try preempt in `default-flavor1`. If preemptor return preempt fail, we try to allocate resource in `default-flavor2`.
+
+We will pass preemptor to flavorAssigner and let flavorAssigner try to call public interface of preemptor to get candidate for certain podSet in current flavor if podSet can not find enough resources in current flavor.
```

Comment:
How will the preemptor know  which flavors are wanted for each call?

Just as a note:  currently , the preemptor depends on flavorassigner , not the other way around.

Reply:
currently preemptor will find resource which need preemption in Do func:

https://github.com/kubernetes-sigs/kueue/blob/e2cb8de09971359fb267dcca61d0eceb8863306b/pkg/scheduler/preemption/preemption.go#L77-L81

resPerFlv contains flavor info and resource name, we can modify preemptor to allow preemptor accept resPerFlv as argument directly.

we can can achieve this by moving `resourcesRequiringPreemption` and `totalRequestsForAssignment` to outside.

Reply:
Please update based on #806